crates/bevy_diagnostic/src/diagnostic.rs

```diff
@@ -227,24 +236,70 @@ impl Diagnostics {
             .and_then(|diagnostic| diagnostic.measurement())
     }
 
+    /// Return an iterator over all [`Diagnostic`].
+    pub fn iter(&self) -> impl Iterator<Item = &Diagnostic> {
+        self.diagnostics.values()
+    }
+}
+
+#[derive(SystemParam)]
+pub struct Diagnostics<'w, 's> {
+    store: Res<'w, DiagnosticsStore>,
+    queue: Deferred<'s, DiagnosticsBuffer>,
+}
+
+impl<'w, 's> Diagnostics<'w, 's> {
     /// Add a measurement to an enabled [`Diagnostic`]. The measurement is passed as a function so that
     /// it will be evaluated only if the [`Diagnostic`] is enabled. This can be useful if the value is
     /// costly to calculate.
     pub fn add_measurement<F>(&mut self, id: DiagnosticId, value: F)
     where
         F: FnOnce() -> f64,
     {
-        if let Some(diagnostic) = self
-            .diagnostics
-            .get_mut(&id)
+        if self
+            .store
+            .get(id)
             .filter(|diagnostic| diagnostic.is_enabled)
+            .is_some()
         {
-            diagnostic.add_measurement(value());
+            let measurement = DiagnosticMeasurement {
+                time: Instant::now(),
+                value: value(),
+            };
+            self.queue.0.insert(id, measurement);
         }
     }
+}
 
-    /// Return an iterator over all [`Diagnostic`].
-    pub fn iter(&self) -> impl Iterator<Item = &Diagnostic> {
-        self.diagnostics.values()
+#[derive(Default)]
+struct DiagnosticsBuffer(StableHashMap<DiagnosticId, DiagnosticMeasurement>);
+
+impl SystemBuffer for DiagnosticsBuffer {
+    fn apply(
+        &mut self,
+        _system_meta: &bevy_ecs::system::SystemMeta,
+        world: &mut bevy_ecs::world::World,
+    ) {
+        let mut diagnostics = world.resource_mut::<DiagnosticsStore>();
+        for (id, measurement) in self.0.drain() {
+            if let Some(diagnostic) = diagnostics.get_mut(id) {
+                diagnostic.add_measurement(measurement);
+            }
+        }
+    }
+}
+
+/// Easy way to register a new diagnostic with an App.
+/// This could equally be a native fn of the app behind a feature flag.
```

Comment:
```suggestion
/// Register a new [`Diagnostic`] with an [`App`].
```

I suggest keeping to the point in doc strings. It might also benefit from a small code example, but not necessary.